README.md
[![Build Status](https://travis-ci.org/LaDilettante/vietnamcode.svg?branch=master)](https://travis-ci.org/LaDilettante/vietnamcode)
[![codecov](https://codecov.io/gh/LaDilettante/vietnamcode/branch/master/graph/badge.svg)](https://codecov.io/gh/LaDilettante/vietnamcode)
[![CRAN](http://www.r-pkg.org/badges/version/vietnamcode)](http://www.r-pkg.org/badges/version/vietnamcode)

# Usage

Use the main function `vietnamcode` to convert Vietnam's provinces' names and ID across different formats. The function can handle diacritics and various spellings.

```
source <- "HCMC"
vietnamcode::vietnamcode(source, "province_name", "province_name")
vietnamcode::vietnamcode(source, "province_name", "pci")
vietnamcode::vietnamcode(source, "province_name", "enterprise_census")
```

# Installation

```
install.packages("vietnamcode") # Or
devtools::install_github("LaDilettante/vietnamcode")
```


NEWS.md
# vietnamcode 0.1.1

Use custom function `utils/convert_unicode_to_ascii` instead of relying on `iconv(s, to="ASCII//TRANSLIT")`. The package is now portable across OSes, especially Windows.

# vietnamcode 0.1.0

First release


tests/testthat.R
library(testthat)
library(vietnamcode)

test_check("vietnamcode")


tests/testthat/test-vietnamcode.R
context("Province name conversion")

test_that("Edge case for number of strings", {
  expect_equal(vietnamcode(c("\u00d0\u00e0 N\u1eb5ng"), "province_name", "province_name"),
               "Da Nang")
})

test_that("Returns non-match as NA", {
  expect_equal(vietnamcode(c(""), "province_name", "pci"), NA_character_)
  expect_equal(vietnamcode(c("haha"), "province_name", "pci"), NA_character_)
  expect_equal(vietnamcode(c("haha", "b"), "province_name", "pci"),
               c(NA_character_, NA_character_))
})

test_that("Handles null argument", {
  expect_error(vietnamcode(c(), "pci", "enterprise_census"), "sourcevar cannot be NULL")
  expect_error(vietnamcode(NULL, "pci", "enterprise_census"), "sourcevar cannot be NULL")
})

test_that("Handles province name with diacritics", {
  expect_equal(vietnamcode(c("\u00d0\u00e0 N\u1eb5ng"), "province_name", "province_name"),
              c("Da Nang"))
  expect_equal(vietnamcode(c("B\u1eafc K\u1ea1n", "S\u00f3c Trang" ), "province_name", "province_name"),
               c("Bac Kan", "Soc Trang"))
})

test_that("Handles lower / upper case", {
  expect_equal(vietnamcode(c("hA nOI"), "province_name", "province_name"),
               c("Ha Noi"))
})

test_that("Search province name using regex", {
  expect_equal(vietnamcode(c("HCMC", "TP Ho Chi Minh", "TP. Ho Chi Minh"),
                           "province_name", "province_name"),
               rep("TP HCM", 3))
  expect_equal(vietnamcode(c("BR.VT", "BR-VT", "Ba Ria-Vung Tau"),
                           "province_name", "province_name"),
               rep("BRVT", 3))
})

test_that("Handles PCI code", {
  expect_equal(vietnamcode(c("1", 2, "42"), "pci", "province_name"),
               c("Ha Noi", "Hai Phong", "Ha Tay"))
})


tests/testthat/test-unicode-conversion.R
context("Convert Vietnamese Unicode to ASCII")

test_that("Converts Unicode to ASCII", {
  expect_equal(convert_unicode_to_ascii("\u00C0"), "A")
  expect_equal(convert_unicode_to_ascii("\u1EF9"), "y")
  expect_equal(convert_unicode_to_ascii("\u00d0"), "D")
})


R/utils.R
which_or_NA <- function(v) {
  ind <- which(v)
  if (length(ind) == 0) {ind <- NA_integer_}
  return(ind)
}

convert_unicode_to_ascii <- function(string) {
  old <- paste0("\u00C0\u00C1\u00C2\u00C3\u00C8\u00C9\u00CA\u00CC\u00CD\u00D2",
                "\u00D3\u00D4\u00D5\u00D9\u00DA\u00DD\u00E0\u00E1\u00E2\u00E3",
                "\u00E8\u00E9\u00EA\u00EC\u00ED\u00F2\u00F3\u00F4\u00F5\u00F9",
                "\u00FA\u00FD\u0102\u0103\u0110\u0111\u0128\u0129\u0168\u0169",
                "\u01A0\u01A1\u01AF\u01B0\u1EA0\u1EA1\u1EA2\u1EA3\u1EA4\u1EA5",
                "\u1EA6\u1EA7\u1EA8\u1EA9\u1EAA\u1EAB\u1EAC\u1EAD\u1EAE\u1EAF",
                "\u1EB0\u1EB1\u1EB2\u1EB3\u1EB4\u1EB5\u1EB6\u1EB7\u1EB8\u1EB9",
                "\u1EBA\u1EBB\u1EBC\u1EBD\u1EBE\u1EBF\u1EC0\u1EC1\u1EC2\u1EC3",
                "\u1EC4\u1EC5\u1EC6\u1EC7\u1EC8\u1EC9\u1ECA\u1ECB\u1ECC\u1ECD",
                "\u1ECE\u1ECF\u1ED0\u1ED1\u1ED2\u1ED3\u1ED4\u1ED5\u1ED6\u1ED7",
                "\u1ED8\u1ED9\u1EDA\u1EDB\u1EDC\u1EDD\u1EDE\u1EDF\u1EE0\u1EE1",
                "\u1EE2\u1EE3\u1EE4\u1EE5\u1EE6\u1EE7\u1EE8\u1EE9\u1EEA\u1EEB",
                "\u1EEC\u1EED\u1EEE\u1EEF\u1EF0\u1EF1\u1EF2\u1EF3\u1EF4\u1EF5",
                "\u1EF6\u1EF7\u1EF8\u1EF9\u00d0")
  new <- paste0("AAAAEEEIIO",
                "OOOUUYaaaa",
                "eeeiioooou",
                "uyAaDdIIUu",
                "OoUuAaAaAa",
                "AaAaAaAaAa",
                "AaAaAaAaEe",
                "EeEeEeEeEe",
                "EeEeIiIiOo",
                "OoOoOoOoOo",
                "OoOoOoOoOo",
                "OoUuUuUuUu",
                "UuUuUuYyYy",
                "YyYyD")
  return(chartr(old, new, string))
}


R/vietnamcode.R
#' Convert Vietnam provincial ID
#'
#' Converts Vietnam's provinces' names and ID across different formats. Handles
#' diacritics and different spellings.
#'
#' @param sourcevar Character vector that contains the codes or province names to be converted
#' @param origin String that indicates the coding scheme of origin
#' @param destination String that indicates the coding scheme of destination
#'
#' @keywords vietnamcode
#' @note Supports the following coding schemes:
#' \itemize{
#'  \item{province_name}
#'  \item{province_name_diacritics}
#'  \item{enterprise_censusm enterprise_census_old, enterprise_census_c: }{the same as GSO code}
#'  \item{pci: }{Provincial Competitiveness Index}
#' }
#'
#' @export
#' @examples
#' codes.of.origin <- vietnamcode::vietnamcode_data$province_name # Vector of values to be converted
#' vietnamcode(codes.of.origin, "province_name", "pci")
vietnamcode <- function(sourcevar,
  origin = c("province_name",
             "enterprise_census", "enterprise_census_old", "enterprise_census_c",
             "pci"),
  destination = c("province_name", "province_name_diacritics",
                "enterprise_census", "enterprise_census_old", "enterprise_census_c",
                "pci")) {
  if (is.null(sourcevar))
    stop("sourcevar cannot be NULL")

  origin <- match.arg(origin)
  destination <- match.arg(destination)

  # Sanitize province name to lower case and ASCII
  if (origin %in% c("province_name")) {
    sourcevar <- tolower(convert_unicode_to_ascii(sourcevar))

    # Multiple regex search
    tmp <- sapply(stats::na.omit(vietnamcode::vietnamcode_data[["regex"]]), grepl,
           unique(sourcevar), ignore.case = TRUE, perl = TRUE)
    if (is.vector(tmp)) {
      regex_index = which_or_NA(tmp)
    } else if (is.matrix(tmp)) {
      regex_index = apply(tmp, 1, which_or_NA)
    }

    match_table <- data.frame(source = unique(sourcevar),
                              regex_index = regex_index)

    destination_index <- match_table[match(sourcevar, match_table[["source"]]),
                                     "regex_index"]
  } else {
    destination_index <- match(sourcevar, vietnamcode::vietnamcode_data[[origin]])
  }

  return(vietnamcode::vietnamcode_data[destination_index, destination])
}


data-raw/pci_gso.R
library(haven)
library(plyr)
library(dplyr)

d_census <- read_dta("data/province_codes.dta")
d_census <- d_census %>% arrange(province) %>%
  select(province_name = province,
         enterprise_census = tinh_no,
         enterprise_census_old = census_old) %>%
  mutate(province_name = ifelse(enterprise_census %in% c(97, 98, 99),
                                NA, province_name)) %>%
  mutate(province_name = ifelse(enterprise_census == 44,
                                "Quang Binh", province_name)) %>%
  mutate(province_name = revalue(province_name,
                                 c("Bac Can" = "Bac Kan",
                                   "Hanoi" = "Ha Noi", "HCMC" = "TP HCM",
                                   "TT-Hue" = "Thua Thien Hue")))

d_pci <- read_dta("data-raw/PCI2011.dta") %>%
  select(pci = pci_id)
d_pci = data.frame(pci = attr(d_pci$pci, "labels")) %>%
  mutate(province_name = row.names(.)) %>%
  mutate(province_name = revalue(province_name,
                                 c("HCMC" = "TP HCM",
                                   "TT-Hue" = "Thua Thien Hue"))) %>%
  rbind(c(42, "Ha Tay")) # According to PCI2005

d_name <- read.csv("data-raw/regex.csv", stringsAsFactors = FALSE)

vietnamcode_data <- full_join(d_census, d_name, by = "province_name") %>%
  full_join(d_pci, by = "province_name") %>%
  mutate(enterprise_census_c = formatC(enterprise_census,
                                       width = 2, format = "d", flag = "0"))

write.csv(vietnamcode_data, file = "data/vietnamcode_data.csv",
          row.names = FALSE)
devtools::use_data(vietnamcode_data, overwrite = TRUE)


data-raw/create_vietnamcode_data.R
library(haven)
library(plyr)
library(dplyr)

d_census <- read_dta("data-raw/province_codes.dta")
d_census <- d_census %>% arrange(province) %>%
  select(province_name = province,
         enterprise_census = tinh_no,
         enterprise_census_old = census_old) %>%
  mutate(province_name = ifelse(enterprise_census %in% c(97, 98, 99),
                                NA, province_name)) %>%
  mutate(province_name = ifelse(enterprise_census == 44,
                                "Quang Binh", province_name)) %>%
  mutate(province_name = revalue(province_name,
                                 c("Bac Can" = "Bac Kan",
                                   "Hanoi" = "Ha Noi", "HCMC" = "TP HCM",
                                   "TT-Hue" = "Thua Thien Hue")))

d_pci <- read_dta("data-raw/PCI2011.dta") %>%
  select(pci = pci_id)
d_pci = data.frame(pci = attr(d_pci$pci, "labels")) %>%
  mutate(province_name = row.names(.)) %>%
  mutate(province_name = revalue(province_name,
                                 c("HCMC" = "TP HCM",
                                   "TT-Hue" = "Thua Thien Hue"))) %>%
  rbind(c(42, "Ha Tay")) # According to PCI2005

d_name <- read.csv("data-raw/regex.csv",
                   fileEncoding = "UTF-8", stringsAsFactors = FALSE)

vietnamcode_data <- full_join(d_census, d_name, by = "province_name") %>%
  full_join(d_pci, by = "province_name") %>%
  mutate(enterprise_census_c = formatC(enterprise_census,
                                       width = 2, format = "d", flag = "0")) %>%
  mutate_each(funs(as.character))

Encoding(vietnamcode_data$province_name_diacritics) <- "UTF-8"
class(vietnamcode_data) <- "data.frame"

write.csv(vietnamcode_data, file = "data/vietnamcode_data.csv",
          row.names = FALSE)
devtools::use_data(vietnamcode_data, overwrite = TRUE)
